File: executors/node/likely_subtags.js
// Tests Intl Locale for minimize / maximize likely subtags.

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale


module.exports = {
  testLikelySubtags: function (json) {
    const label = json['label'];
    const locale = json['locale'];
    let test_option;
    if (json['option']) {
      test_option = json['option'];
    }

    let return_json = {'label': label};
    let intl_locale;
    try {
      intl_locale = new Intl.Locale(locale);
    } catch (error) {
      /* Something is wrong with the constructor */
      return_json['error'] = 'CONSTRUCTOR: ' + error.message;
      return return_json;
    }

    try {
      let result_locale;
      if (test_option === 'maximize') {
        result_locale = intl_locale.maximize().baseName;
      } else if (test_option === 'minimizeFavorRegion' ||
                 test_option === 'minimize') {
        result_locale = intl_locale.minimize().baseName;
        // Unlikely subtags: lang is "und", result is same as input, and favor region
        if (result_locale == locale && locale.split('-')[0] == 'und' &&
           test_option == 'minimizeFavorRegion') {
          // TODO: set unsupported as function taking test option and unsupported field.
          return_json['error_detail'] = test_option;
          return_json['error_type'] = 'unsupported';
          return_json['unsupported'] = 'UND not supported with these options';
        }
      } else {
        return_json['error_detail'] = test_option;
        return_json['error_type'] = 'unsupported';
        return_json['unsupported'] = 'Unknown test option';
      }
      return_json['result'] = result_locale;
    } catch (error) {
      return_json['unsupported'] = 'LIKELY_SUBTAGS UNKNOWN ERROR: ' + error.message;
    }
    return return_json;
  }

}

File: executors/rust/src/likelysubtags.rs
//! Provides tests for likely subtags to mimimize and maximize.

use serde_json::{json, Value};

#[cfg(any(ver = "1.3", ver = "1.4", ver = "1.5"))]
pub use icu::locid_transform::LocaleExpander;

#[cfg(not(any(ver = "1.3", ver = "1.4", ver = "1.5")))]
pub use icu::locale::LocaleExpander;

#[cfg(any(ver = "1.3", ver = "1.4", ver = "1.5"))]
type LocaleType = super::compat::Locale;

#[cfg(not(any(ver = "1.3", ver = "1.4", ver = "1.5")))]
type LocaleType = super::compat::LanguageIdentifier;

// https://docs.rs/icu/latest/icu/locid_transform/

// Function runs language names tests
pub fn run_likelysubtags_test(json_obj: &Value) -> Result<Value, String> {
    let lc = LocaleExpander::new_extended();

    let label = &json_obj["label"].as_str().unwrap();

    let test_option = &json_obj["option"].as_str().unwrap();

    let locale_str: &str = json_obj["locale"].as_str().unwrap();

    let mut locale = locale_str.parse::<LocaleType>().unwrap();

    if test_option == &"minimizeFavorScript" {
        #[cfg(any(ver = "1.3", ver = "1.4"))]
        {
            // This option is not yet supported.
            return Ok(json!({
                "label": label,
                "error_detail": {"option": test_option},
                "unsupported": test_option,
                "error_type": "unsupported",
            }));
        }
        #[cfg(not(any(ver = "1.3", ver = "1.4")))]
        {
            lc.minimize_favor_script(&mut locale);
        }
    } else if test_option == &"minimize" || test_option == &"minimizeFavorRegion" {
        lc.minimize(&mut locale);
    } else if test_option == &"maximize" {
        lc.maximize(&mut locale);
    } else {
        // This option is not yet supported.
        return Ok(json!({
            "label": label,
            "error_detail": {"option": test_option},
            "unsupported": test_option,
            "error_type": "unsupported",
        }));
    }

    let json_result = json!({
        "label": label,
        "result": locale.to_string()
    });

    Ok(json_result)
}
